pkg/apis/traceflow/v1/types.go

```diff
@@ -0,0 +1,162 @@
+	DROPPED   Action = "DROPPED"
+)
+
+type PacketResouceType string
+
+const (
+	FieldsPacketData PacketResouceType = "FieldsPacketData"
+)
+
+type PacketTransportType string
+
+const (
+	UNICAST PacketTransportType = "UNICAST"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Source      `json:"source"`
+	Destination `json:"destination"`
+	RoundID     string `json:"roundID,omitempty"`
+
+	Packet `json:"packet"`
+	Status `json:"status"`
+}
+
+type Source struct {
+	SrcNamespace string `json:"srcNamespace,omitempty"`
+	SrcPod       string `json:"srcPod,omitempty"`
+}
+
+type Destination struct {
+	DstNamespace string `json:"dstNamespace,omitempty"`
+	DstPod       string `json:"dstPod,omitempty"`
+	DstService   string `json:"dstService,omitempty"`
+	DstIP        string `json:"dstIP,omitempty"`
+}
+
+type IPHeader struct {
+	SrcIP    string `json:"srcIP,omitempty"`
+	Protocol int32  `json:"protocol,omitempty"`
+	TTL      int32  `json:"ttl,omitempty"`
+	Flags    int32  `json:"flags,omitempty"`
+}
+
+type TransportHeader struct {
+	ICMPEchoRequestHeader `json:"icmpEchoRequestHeader"`
+	UDPHeader             `json:"udpHeader"`
+	TCPHeader             `json:"tcpHeader"`
+}
+
+type ICMPEchoRequestHeader struct {
+	ID       int32 `json:"id,omitempty"`
+	Sequence int32 `json:"sequence,omitempty"`
+}
+
+type UDPHeader struct {
+	SrcUDPPort int32 `json:"srcUDPPort,omitempty"`
+	DstUDPPort int32 `json:"dstUDPPort,omitempty"`
+}
+
+type TCPHeader struct {
+	SrcTCPPort int32 `json:"srcTCPPort,omitempty"`
+	DstTCPPort int32 `json:"dstTCPPort,omitempty"`
+	TCPFlags   int32 `json:"tcpFlags,omitempty"`
+}
+
+type Packet struct {
+	ResouceType   PacketResouceType   `json:"resouceType,omitempty"`
+	TransportType PacketTransportType `json:"transportType,omitempty"`
+	PayloadString string              `json:"payloadString,omitempty"`
+	PayloadSize   int32               `json:"payloadSize,omitempty"`
+
+	IPHeader        `json:"ipHeader"`
+	TransportHeader `json:"transportHeader"`
+}
+
+type Status struct {
+	Phase        Phase         `json:"phase,omitempty"`
+	CrossNodeTag uint8         `json:"crossNodeTag,omitempty"`
+	NodeSender   []Observation `json:"nodeSender,omitempty"`
+	NodeReceiver []Observation `json:"nodeReceiver,omitempty"`
+}
+
+type Observation struct {
+	Component       Component  `json:"component,omitempty"`
+	SubComponent    string     `json:"subComponent,omitempty"`
+	ComponentName   string     `json:"componentName,omitempty"`
+	Action          Action     `json:"action,omitempty"`
+	RoundID         string     `json:"roundID,omitempty"`
+	NodeUUID        string     `json:"nodeUUID,omitempty"`
+	PodUUID         string     `json:"podUUID,omitempty"`
+	DstMAC          string     `json:"dstMAC,omitempty"`
+	RuleID          string     `json:"ruleID,omitempty"`
```

Comment:
Do we know what will be filled into Rule and what will be RuleID? If we are not clear of a field yet, maybe remove it for now.

Reply:
RuleID=conjunctionID
Rule is not clear now.

Reply:
I think we should use an ID that can be calculated deterministically from the rule expression, so the client can look up the rule from the ID.
@tnqn : do you have an idea?

Reply:
Do you mean NetworkPolicy which you commented below?
I'm investigating if it's possible to get network policy from conjunctionID on antrea-agent.
@tnqn Any idea?

Reply:
By "deterministically" I think you mean hashing? The conjunction ID is a 32-bit value, it's quite possible to have collision and it doesn't help look up the rule from the ID. 
I think we need a reverse index anyway to look up rule by ID and it's not hard to add. I can add an interface for that. 
Should we use the policy namespaced name here? If it's other information like UUID, user still cannot associate it with policy directly.

Reply:
We just need to look up the rule from a NetworkPolicy, and the chance of collision should be pretty low?
Basically I want some ID that can be used by a client/user to find out the exact rule from the NetworkPolicy. I am thinking about the following options: 1) the order of the rules in the rule slice; 2) hash of the rule struct; 3) an ID generated by controller and sent to agent. What you think?

Reply:
@tnqn I found we have a private function named "getPolicyRuleConjunction", which can query "policyRuleConjunction" by conjunction ID. NetworkPolicy name and Namespace information is included in "policyRuleConjunction" for debugging usage.
Can we make a public interface to query NetworkPolicy name and Namespace from here?
My usage: https://github.com/vmware-tanzu/antrea/pull/731/commits/bdbb8b8742dc3def27892479900142dc55ae0dde#diff-656cf5e26def0c66837e98477c6ed2ba

Reply:
@jianjuns then the "ruleID" doesn't mean the conjunction ID, right? If we want it informative to user, I think it should the rule index. But on the other hand, we made some change to the rules when converting K8s networkpolicy to internal one and non rules contribute to the drop action but the policy itself, so I wonder if we should just have networkpolicy here but not ruleID.

Reply:
I think better to have rule ID, not just for Traceflow, but also other features like flow exporting/correlation, realization, stats, etc.

Again could we use some ID that user can use to identify a rule, ideally from K8s API, or at least from Controller API? If that means Controller needs to pass the ID to Agent, we can add it to NP messages?

If this requires too big changes, let us remove it for the current version, and add it later.

Reply:
@jianjuns In current version, we provide ConjuntionID (unser ruleID section) and NetworkPolicy in Observation.
e.g.
```
  - action: FORWARDED
    component: NETWORKPOLICY
    componentName: EGRESS
    networkPolicy: default/allow-all  # This is NetworkPolicy
    node: zhecheng-k8s2-0
    ruleID: "2"                                     # This is ConjunctionID
    timestamp: 1592276104
```

Reply:
I do not feel ConjunctionID is useful, as user has no idea what it is unless it goes to every Node to check.

Reply:
@tnqn How about your opinion to remove ConjunctionID from the result? See my last comment above for example.

Reply:
Agree with @jianjuns that ConjunctionID is not useful to user and we should add ability to identify rules in policy first.

Reply:
Seems we can remove ruleID in this release and add it back if needed.

Reply:
Removed ruleID.

Reply:
@lzhecheng , please file one issue to track "ruleID" in Antrea 0.9.

Reply:
Track `RuleID` in an issue: https://github.com/vmware-tanzu/antrea/issues/845